#!/bin/bash
#$ -l h_rt=24:00:00
#$ -l h_vmem=10G
#$ -o /path/to/Detect_circRNA-$JOB_ID.out
#$ -e /path/to/Detect_circRNA-$JOB_ID.err
#$ -N Detect_for_circRNA
#$ -j y
#$ -q queue
#____________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________

#export your Miniconda path 


#____________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________

# Get the current timestamp:
start_timestamp=$(date "+%Y-%m-%d %H:%M:%S")

# Print the timestamp to stdout:
echo "                                "
echo "                                "
echo "Job started at: $start_timestamp"
echo "                                "


#____________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________
# Activate circtools
source activate circTools

echo "Activated circtools"

#____________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________

#Path to circtools directory containing samplesheets for input
circtools_dir="/path/to/4__Prep_Detect"

#Path to output directory 
Output_Dir="/path/to/A___Detect"

#Path to Ensembl repeats annotation gtf file used for detection (Output of Step 3___Create_Ensembl_Reference)
Repeats_GTF="/path/to/ensembl_repeats.gtf"

# Path to the annotation GTF file:
Annotation_GTF="/path/to/GTF_file.gtf"

#Path to reference genome FASTA file
Genome_FASTA="/path/to/genome_FASTA.fa"

#Path to circtools repository (github clone)
circtools_repo="/path/to/circtools_dir/circtools"

#____________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________

# Run circtools detection to detect circRNAs
python "$circtools_repo/circtools.py" detect @"$circtools_dir/samplesheet" \
    -mt1 @"$circtools_dir/mate1" \
    -mt2 @"$circtools_dir/mate2" \
    -B @"$circtools_dir/bam_files.txt" \
    -D \
    -R "$Repeats_GTF" \
    -an "$Annotation_GTF" \
    -Pi \
    -F \
    -M \
    -Nr 3 5 \
    -fg \
    -A "$Genome_FASTA" \
    -G \
    -L 20 \
    -O "${Output_Dir}/5__Detect_output"



#____________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________

# Memory Used:
echo "                       "
echo "Let's look at the vmem:"
qstat -j $JOB_ID | grep vmem 
echo "                       "

# Get the timestamp when the command completes:
end_timestamp=$(date "+%Y-%m-%d %H:%M:%S")

# _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

# Print the end timestamp to stdout:
echo "                            "
echo "Job ended at: $end_timestamp"
echo "                            "

#____________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________